text/0001-workspace-consistent-versions.md

```diff
@@ -0,0 +1,570 @@
+# pnpm-lock.yaml
+workspaceConsistent:
+  react:
+    specifier: ^17.0.0
+    dependency: 17.0.2
+```
+
+Suppose semver ranges were indeed allowed on `workspace-consistent:`. If a package were to specify `workspace-consistent:^`:
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "workspace-consistent:^",
+  },
+}
+```
+
+On publish this is expected to become:
+
+```json5
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "^17.0.0",
+  },
+}
+```
+
+It would be surprising if this became `^17.0.2` (created by adding `^` to the resolved version) since that may be overly specific and leave authors with no way to relax that.
+
+On the other hand, suppose a package were to specify `workspace-consistent:*`:
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "workspace-consistent:*",
+  },
+}
+```
+
+On publish, the least suprising behavior would be for this to become `17.0.2`. This is because `17.0.2` was the resolved version actually tested on in the monorepo.
+
+```json5
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "17.0.2",
+  },
+}
+```
+
+This leads to a discrepancy between what `workspace-consistent:*` and `workspace-consistent:^` should append semver range types to on publish. (The original specifier vs the resolved version.)
+
+The semantics may become more clear if `workspaceConsistent` configured versions must be a concrete version, but this eliminates the ability to `pnpm update` `workspaceConsistent` values.
+
+### Lockfile `workspaceConsistent` block format
+
+The proposed format is:
+
+```yaml
+workspaceConsistent:
+  jest:
+    specifier: ^27.5.1
+    dependency: 27.5.1
+
+  react:
+    specifier: ^17.0.2
+    dependency: 17.0.2
+
+  react-dom:
+    specifier: ^17.0.2
+    dependency: 17.0.2
+```
+
+An alternative format that represents the same information would be:
+
+```yaml
+workspaceConsistent:
+  specifiers:
+    jest: ^27.5.1
+    react: ^17.0.2
+    react-dom: ^17.0.2
+  dependencies:
+    jest: 27.5.1
+    react: 17.0.2
+    react-dom: 17.0.2
+```
+
+The alternative format better matches the existing shape of `importers`. However the proposed format is more resilient to merge conflicts. Separate commits changing the version of `react` and `jest` would merge conflict with the alternative format, while the proposed format would merge those changes cleanly.
+
+## Extensions
+
+### Transitive Consistency
+
+Multiple versions of a package can still end up in the final product due to versions declared as a dependency of a dependency.
+
+It would be useful to declare two extra fields to solve the duplicate dependency problem completely.
+
+```json5
+{
+  "name": "example-monorepo",
+  "private": true,
+  "pnpm": {
+    "workspaceConsistent": {
+      "groups": {
+        "default": {
+          "@types/react": "^17.0.43",
+          "react": "^17.0.2",
+          "react-dom": "^17.0.2",
```

Comment:
For publishing, I think it would be most useful if the definition of the version group defined both the exact version that is to be installed in the workspace, and separately the specifier that is to be written during publishing. This will allow developers to loosen the version for their customers while still controlling exactly what gets installed in the workspace (which is increasingly necessary in the current security landscape).

Reply:
Open to this, but want to ask a few more questions.

1. Does specifying both a range specifier for internal upgrades and publishing run into the scenario [described here](https://github.com/pnpm/rfcs/blob/9baf724550b6691f5171fc963da7c5d77057fba2/text/0001-workspace-consistent-versions.md#semver-range-specifier-in-workspaces)? Namely:
   - If you have an internal upgrade range of `^17.0.0`, and a resolved internal lockfile version of `17.0.1` what should a publish range of `*` result in? `17.0.0` or `17.0.1`?
   - In a similar scenario but a publish range of `^`, should the dependency publish with `^17.0.0` or `^17.0.1`? I think the answer to this scenario is more clearly `^17.0.0`, but it results in confusing behavior viewed next to the ambiguous behavior for `*`.

2. I might be missing something, but I think this is the existing situation for how dependencies are declared and resolved at the moment too? Is there something special about workspace consistent versions that means this feature should do something additional?
   - To elaborate, if someone wants to keep the version range for customers loose while controlling exactly what gets installed into the workspace, they have to modify the `pnpm-lock.yaml` file (either manually or through a command).

Reply:
I'm a bit confused by your linked example, in that I'd expect the only data in package.json to be `"react": "workspace-consistent:react-rc"`. The version range is specified in the definition of `"react-rc"`. I'd like to be able to do something like:
```
"react-rc": {
  "react": {
    "publish": "^17",
    "install": "17.0.2"
  }
}
```
```
"dependencies": {
  // Resolves to "17.0.2" at install time, when published is "^17"
  "react": "workspace-consistent:react-rc"
}
```
The value in "publish" is a superset of the value in "install". Technically you could do the same thing by declaring two dependency sets "react-rc-dev" and "react-rc-publish" and having the former as `devDependencies` and the latter as regular `dependencies`, since `devDependencies` supersede regular `dependencies` in local packages, but that's rather cumbersome compared to declaring it inside the definition of the version group.

Reply:
Apologies for the late response. I do appreciate your input.

Are there downsides to keeping the resolved version in `pnpm-lock.yaml`? Adapting your example:

```json
"react-rc": {
  "react": "^17"
}
```

```yaml
# pnpm-lock.yaml

workspaceConsistent:
  react-rc:
    react:
      specifier: ^17
      dependency: 17.0.2
```

Although it may not be intuitive, it's possible to edit the `dependency: 17.0.2` line in `pnpm-lock.yaml` instead of editing an `"install": "17.0.2"` field in `package.json`. I think this accomplishes the same thing?

The benefits of keeping resolved versions in `pnpm-lock.yaml` are:

- Follows the current convention of keeping concrete/resolved versions in `pnpm-lock.yaml`.
- `pnpm update` doesn't have to edit the root `package.json`, which I think can become strange. Admittedly I haven't tested this command before, but from the [description in the docs](https://pnpm.io/cli/update) I'd imagine it only edits `pnpm-lock.yaml` at the moment.
- Users don't have to manually look up or decide concrete versions on npmjs.com when adding a new dependency.

To be fair, I think the benefits of having resolved versions in `package.json` are:

- It's apparent to users that you can edit the resolved version in `package.json`. (Editing `pnpm-lock.yaml` is probably discouraged on most teams.)

Is everything above correct? Did I misunderstand the pros/cons?